From 470c9354a19626944d3ec4bc3a06d99b8fac4970 Mon Sep 17 00:00:00 2001
From: Lothar Rubusch <lothar.rubusch@enclustra.com>
Date: Thu, 12 Sep 2024 17:59:15 +0000
Subject: [PATCH] drivers: atsha204a fix something

repatched:

From 13d6d12502519c44710e09183323feb6e37c252b Mon Sep 17 00:00:00 2001
From: Andreas Buerkler <andreas.buerkler@enclustra.com>
Date: Wed, 7 Dec 2022 11:37:18 +0100
Subject: [PATCH] bugfix for atsha204a driver

Signed-off-by: Lothar Rubusch <lothar.rubusch@enclustra.com>
---
 drivers/misc/atsha204a-i2c.c | 35 ++++++++++++++++++++++++-----------
 1 file changed, 24 insertions(+), 11 deletions(-)

diff --git a/drivers/misc/atsha204a-i2c.c b/drivers/misc/atsha204a-i2c.c
index d3c515828f..549f16da80 100644
--- a/drivers/misc/atsha204a-i2c.c
+++ b/drivers/misc/atsha204a-i2c.c
@@ -94,29 +94,38 @@ static int atsha204a_recv_resp(struct udevice *dev,
 
 int atsha204a_wakeup(struct udevice *dev)
 {
-	u8 req[4];
 	struct atsha204a_resp resp;
 	int try, res;
 
 	debug("Waking up ATSHA204A\n");
 
-	for (try = 1; try <= 10; ++try) {
+	for (try = 1; try <= 5; ++try) {
 		debug("Try %i... ", try);
 
-		/*
-		 * The device ignores any levels or transitions on the SCL pin
-		 * when the device is idle, asleep or during waking up.
-		 * Don't check for error when waking up the device.
-		 */
-		memset(req, 0, 4);
-		atsha204a_send(dev, req, 4);
+		// Send 1 byte zero to wakeup the device.
+		// The I2C frequency must be below 133kHz
+		// to generate a low pulse longer than the required 60us.
+		struct i2c_msg msg;
+		msg.addr = 0;
+		msg.flags = I2C_M_STOP;
+		msg.len = 0;
+		msg.buf = NULL;
+
+		dm_i2c_xfer(dev, &msg, 1);
+
+		// Test if the device is awake
+		fdt_addr_t *priv = dev_get_priv(dev);
+		msg.addr = *priv;
+		res = dm_i2c_xfer(dev, &msg, 1);
+		if (res)
+			continue;
 
 		udelay(ATSHA204A_TWLO_US + ATSHA204A_TWHI_US);
 
 		res = atsha204a_recv_resp(dev, &resp);
 		if (res) {
-			debug("failed on receiving response, ending\n");
-			return res;
+			debug("failed on receiving response\n");
+			continue;
 		}
 
 		if (resp.code != ATSHA204A_STATUS_AFTER_WAKE) {
@@ -274,7 +283,11 @@ static int atsha204a_of_to_plat(struct udevice *dev)
 	fdt_addr_t *priv = dev_get_priv(dev);
 	fdt_addr_t addr;
 
+	/*
 	addr = dev_read_addr(dev);
+	/*/ // TODO uncomment the above if it works out
+	addr = fdtdec_get_addr_size_auto_noparent(gd->fdt_blob, dev_of_offset(dev), "reg", 0, NULL, false);
+	// */
 	if (addr == FDT_ADDR_T_NONE) {
 		debug("Can't get ATSHA204A I2C base address\n");
 		return -ENXIO;
-- 
2.37.6

